foyer/general_forcefield.py

```diff
@@ -0,0 +1,507 @@
+
+# Copy from original forcefield.py
+def preprocess_forcefield_files(forcefield_files=None, backend='gmso'):
+    """Pre-process foyer Forcefield XML files"""
+    if forcefield_files is None:
+        return None
+
+    tmp_processed_files = list()
+    if backend == 'gmso':
+        # Run through the forcefield XML conversion
+        from gmso.external.convert_foyer_xml import from_foyer_xml
+        for idx, file in enumerate(forcefield_files):
+            _, suffix = os.path.split(file)
+            tempfile = NamedTemporaryFile(suffix=suffix, delete=False)
+            from_foyer_xml(foyer_xml=str(file),
+                          gmso_xml=str(tempfile.name), overwrite=True)
+            tmp_processed_files.append(tempfile.name)
+
+    else:
+        raise FoyerError('Backend not supported')
+
+    return tmp_processed_files
+
+
+class Forcefield(object):
+    """General Forcefield object that can be created by either a GMSO XML
+    forcefield file of Foyer XML forcefield file
+
+    Parameters
+    ----------
+    forcefield_files : list of str, optional, default=None
+        List of forcefield files to load.
+    name : str, optional, None
+        Name of a forcefield to load that is packaged within foyer.
+    backend : str, optional, default='gmso'
+        Name of the backend used to store all the Types' information.
+        At this point, 'gmso' is the only valid backend, but this set up
+        allow future backend to be implemented easier.
+    """
+    def __init__(self, forcefield_files=None, name=None,
+                       backend='gmso',
+                       **kwargs):
+        self.atomTypeDefinitions = dict()
+        self.atomTypeOverrides = dict()
+        self.atomTypeDesc = dict()
+        self.atomTypeRefs = dict()
+        self.atomTypeClasses = dict()
+        self.atomTypeElements = dict()
+        self._included_forcefields = dict()
+        self.non_element_types = dict()
+        self._version = None
+        self._name = None
+
+        all_files_to_load = []
+        if forcefield_files is not None:
+            if isinstance(forcefield_files, (list, tuple, set)):
+                all_files_to_load = list(forcefield_files)
+            else:
+                all_files_to_load = [forcefield_files]
+
+        if name is not None:
+            try:
+                file = self.included_forcefields[name]
+            except KeyError:
+                raise IOError('Forcefield {} cannot be found.'.format(name))
+            else:
+                all_files_to_load = [file]
+
+        # Preprocessed the input files
+        preprocessed_files = preprocess_forcefield_files(
+                                all_files_to_load,
+                                backend=backend)
+
+        # Load in an internal forcefield object depends on given backend
+        if backend == 'gmso':
+            self._parse_gmso(preprocessed_files, **kwargs)
+        else:
+            raise FoyerError('Backend not supported')
```

Comment:
we should add a test that passes a nonsense backend type and make sure it fails.